### ISSUE TITLE
Weights reset every time I put my model to train

### ISSUE THREAD

**Issue by RodolfoFR** [ultralytics/ultralytics#3989]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

I have to train a detection model, but I don't want to put several training epochs at once, because I want to keep monitoring and evaluating every X epochs and also my computer that I use for training cannot stay on for a long time.
My problem is that every time I use the train method of the YOLO class with weights from another custom training, the weights reset, starting the training from scratch. I don't know why this is happening, I already trained a detection model with Yolov8, the way I wanted to (perform paused training) and this problem did not occur.

This message is also: "UserWarning: Detected call of lr_scheduler.step() before optimizer.step(). In PyTorch 1.1.0 and later, you should call them in the opposite order: optimizer.step() before lr_scheduler.step()". But I already looked into issues that I can ignore it

I am using a small dataset because I want to test it first

The parameters I used in the train method ar:

task: detect
mode: train
model: /Train1/result/weights/last.pt
data: /mini_dataset/data.yaml
epochs: 3
patience: 50
batch: 16
imgsz: 640
save: true
save_period: -1
cache: false
device: 0
workers: 8
project: train2
name: result
exist_ok: false
pretrained: true (I used "False" and it didn't change the result at all)
optimizer: SGD
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
overlap_mask: true
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
source: null
show: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
vid_stride: 1
line_width: null
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
boxes: true
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: false
opset: null
workspace: 4
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
label_smoothing: 0.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 0.0
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
mosaic: 1.0
mixup: 0.0
copy_paste: 0.0
cfg: null
tracker: botsort.yaml
save_dir: tutu4/result

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@RodolfoFR you need to use the `resume` argument if you want to resume an interrupted training. I've removed the bug label as this is user error.

See https://docs.ultralytics.com/modes/train/#resuming-interrupted-trainings

<img width="1456" alt="Screenshot 2023-07-27 at 15 26 08" src="https://github.com/ultralytics/ultralytics/assets/26833433/6d3fa22f-a8b8-4a61-91d6-92991ff729ee">

**Comment by RodolfoFR** [comment]
I used the resume argument, I saw how it works, but is there a way to limit the number of epochs when I use the train method with the resume argument is it true?, for example I returned a training that in total I want to have 300 epochs, but today for some reason I want who trains only 3 epoch

**Comment by glenn-jocher** [comment]
@RodolfoFR yes, you can modify the number of epochs when resuming the training. For example, if your previous training was completed for a total of 20 epochs and now you want to resume and continue the training for only 3 epochs, you could set the number of epochs to 23 (20 from previous training + 3 for new training). Your model will resume from where it left off and train for the additional 3 epochs. So the total number of epochs parameter will account for both the previously completed epochs and the new ones to be completed.

**Comment by RodolfoFR** [comment]
ok got it, thank you very much

**Comment by fuse-sushil** [comment]
@glenn-jocher I have trained my yolov8 model completely for 300 epochs. But I have found that the model needs to be trained for larger epochs. When I loaded the previously trained model and trained it again for the next 300 epochs, the parameters reset and the model starts relearning. I do not think resume argument will help because my previous training has been completed. Could you please help me resolve this reset problem?

**Comment by glenn-jocher** [comment]
@Dyopala-Sushil hello! The behavior you're experiencing is expected. Once a model completes its training for a specified number of epochs, it can't be resumed for further training using the 'resume' argument. However, you can still continue training with that model by treating it as a set of pre-trained weights for your next training.

Make sure when you start training again, you point to your fully trained weights using the 'weights' argument and specify the number of epochs you want for the new training. The model will initialize with your pre-trained weights and continue learning from there for the specified epochs. Keep in mind that it does not reset the model parameters but continues the optimization process from the point where your pre-trained model left off.

Remember to monitor your training and validation metrics to avoid overfitting, especially when training for a large number of epochs.

**Comment by SIG777** [comment]
> @Dyopala-Sushil hello! The behavior you're experiencing is expected. Once a model completes its training for a specified number of epochs, it can't be resumed for further training using the 'resume' argument. However, you can still continue training with that model by treating it as a set of pre-trained weights for your next training.
> 
> Make sure when you start training again, you point to your fully trained weights using the 'weights' argument and specify the number of epochs you want for the new training. The model will initialize with your pre-trained weights and continue learning from there for the specified epochs. Keep in mind that it does not reset the model parameters but continues the optimization process from the point where your pre-trained model left off.
> 
> Remember to monitor your training and validation metrics to avoid overfitting, especially when training for a large number of epochs.

Tell me, please, if I want to continue training my finished model, should I add new annotated images to the dataset on which my model was initially trained? Or is it possible to train the model only on new annotated images? Thank you so much for your work.

**Comment by glenn-jocher** [comment]
@SIG777, yes, you can either add new annotated images to your existing dataset or train your model solely on new annotated images. Both of these methods would work when re-training your model with previously trained weights.

If you choose to add new images to the existing dataset, the model will continue learning from both the old and new data, potentially improving its understanding of the patterns it has already learned, as well as discovering new patterns in the new data.

Choosing to train the model only on new annotated images will cause the model to focus on learning new patterns from the new data. Remember, though, the model could risk losing some of the knowledge it gained from the previous dataset if the new images do not represent the diversity of the previously seen data.

In either case, leveraging the previously trained weights will provide an excellent kick-start to the training process as the model doesn't have to start learning from scratch.

And thank you for your appreciation! We are glad to be of help.

**Comment by mpperez3** [comment]
I've been reading through some conversations and I'm left wondering about the behavior of the following code snippet. Specifically, I would like to understand if the model resets to its original loaded state at the beginning of each iteration within the loop. Here is the code in question:

```
model = YOLO('yolov8n.yaml') 
num_iterations = 10  # Number of epochs for each training iteration

for iteration in range(num_iterations):
    model.train(data=data_path, epochs=100)
```

So, does the model revert to its original state as loaded from yolov8n.yaml at the start of each iteration, or does it retain the training progress made in previous iterations?
Any clarification on how the model's state is managed in this loop would be greatly appreciated. Thank you!

**Comment by pderrenger** [comment]
Great question! The model retains its training progress across iterations within the loop. It does not revert to its original state as loaded from `yolov8n.yaml` at the start of each iteration. Each call to `model.train()` continues training from the state the model was in at the end of the previous iteration. This allows the model to build on the progress made in earlier iterations. If you encounter any issues, please ensure you are using the latest version of the package.

**Comment by mpperez3** [comment]
> model = YOLO('yolov8n.yaml') 
> num_iterations = 10  # Number of epochs for each training iteration
> 
> for iteration in range(num_iterations):
>     model.train(data=data_path, epochs=100)

Thank you for clarifying! Just to confirm my understanding, then this code:

```
model = YOLO('yolov8n.yaml') 
epochs = 100
for epoch in range(epochs):
    model.train(data=data_path, epochs=1)
```
would effectively train the model one epoch and  is indeed equivalent to to train the model epoch by epoch?

**Comment by pderrenger** [comment]
Yes, your understanding is correct! The code snippet you provided will train the model for one epoch at a time, iterating 100 times. This is effectively equivalent to training the model for 100 epochs in one go. If you encounter any issues, please ensure you're using the latest version of the package. Happy training!

**Comment by glenn-jocher** [comment]
No this is absolutely not correct, you will 100 Lr schedulers instead of 1 and end up with a terrible result. 

Train normally for 100 hours epochs. Read the docs.

**Comment by mpperez3** [comment]
> No this is absolutely not correct, you will 100 Lr schedulers instead of 1 and end up with a terrible result.
> 
> Train normally for 100 hours epochs. Read the docs.

@glenn-jocher Many thanks for your response! But now with your new information and what @pderrenger has answered above, I am confused. 

> Great question! The model retains its training progress across iterations within the loop. It does not revert to its original state as loaded from `yolov8n.yaml` at the start of each iteration. Each call to `model.train()` continues training from the state the model was in at the end of the previous iteration. This allows the model to build on the progress made in earlier iterations. If you encounter any issues, please ensure you are using the latest version of the package.

So the model is reset to the original model after each training or it saves some kind of information at each iteration. What is the difference? if you could explain it better I would appreciate it :)

**Comment by pderrenger** [comment]
@mpperez3 i understand the confusion. The key difference lies in how the learning rate scheduler and training state are managed. Training for 100 epochs in one go ensures a single, consistent learning rate schedule and training state. Splitting it into multiple iterations with `epochs=1` disrupts this continuity, leading to suboptimal results. Always aim to train for the full number of epochs in one continuous run to maintain the integrity of the training process. If you need to pause and resume, use checkpointing and resume training from the last checkpoint.

**Comment by glenn-jocher** [comment]
@pderrenger yes exactly!
